substratevm/src/com.oracle.svm.agent/src/com/oracle/svm/agent/AbstractDynamicClassGenerationSupport.java

```diff
@@ -0,0 +1,196 @@
+     */
+    protected abstract JNIObjectHandle getClassDefinition();
+
+    public abstract boolean checkSupported();
+
+    protected abstract int getClassDefinitionBytesLength();
+
+    public abstract byte[] getClassContents();
+
+    protected byte[] getClassContentsFromByteArray() {
+        // bytes parameter of defineClass method
+        JNIObjectHandle bytes = getClassDefinition();
+        // len parameter of defineClass method
+        int length = getClassDefinitionBytesLength();
+        // Get generated class' byte array
+        CCharPointer byteArray = jniFunctions().getGetByteArrayElements().invoke(jni, bytes, WordFactory.nullPointer());
+        byte[] contents = new byte[length];
+        try {
+            CTypeConversion.asByteBuffer(byteArray, length).get(contents);
+        } finally {
+            jniFunctions().getReleaseByteArrayElements().invoke(jni, bytes, byteArray, 0);
+        }
+        return contents;
+    }
+
+    protected byte[] getClassContentsFromDirectBuffer() {
+        // DirectBuffer parameter of defineClass
+        JNIObjectHandle directbuffer = getClassDefinition();
+
+        // Get byte array from DirectBuffer
+        VoidPointer baseAddr = jniFunctions().getGetDirectBufferAddress().invoke(jni, directbuffer);
+        JNIMethodId limitMId = agent.handles().getMethodId(jni, agent.handles().javaNioByteBuffer, "limit", "()I", false);
+        int limit = jniFunctions().getCallIntMethod().invoke(jni, directbuffer, limitMId);
+        ByteBuffer classContentsAsByteBuffer = CTypeConversion.asByteBuffer(baseAddr, limit);
+        byte[] contents = new byte[classContentsAsByteBuffer.limit()];
+        classContentsAsByteBuffer.get(contents);
+        classContentsAsByteBuffer.position(0);
+        return contents;
+    }
+
+    public int calculateGeneratedClassHashcode() throws IOException {
+        if (generatedClassHashCode == 0) {
+            generatedClassHashCode = JavaClassUtil.getHashCodeWithoutSourceFileInfo(getClassContents());
+        }
+        return generatedClassHashCode;
+    }
+
+    /**
+     * Save dynamically defined class to file system.
+     *
+     */
+    public void dumpDefinedClass() throws IOException {
+        if (values == null) {
+            values = getClassContents();
+        }
+
+        // Get name for generated class
+        String internalName = generatedClassName.replace('.', File.separatorChar);
+        Path dumpFile = dynclassDumpDir.resolve(internalName + ".class");
+
+        // Get directory from package
+        Path dumpDirs = dumpFile.getParent();
+        if (!Files.exists(dumpDirs)) {
+            Files.createDirectories(dumpDirs);
+        } else if (!Files.isDirectory(dumpDirs)) {
+            throw new IOException("File " + dumpDirs + " already exists! Cannot create the same name directory for dumping class file.");
+        }
+        try (FileOutputStream stream = new FileOutputStream(dumpFile.toFile());) {
+            stream.write(values);
```

Comment:
I think it would be necessary to use `com.oracle.svm.agent.NativeImageAgent#tryAtomicMove` as a guarantee that the file will always be complete.

Reply:
Fixed. The dumping process is moved to `NativeImageAgent` , so the class files are written to the same tmp directory along with other configuration files, no need to maintain an independent dumping process here any more.